File: lessons/dsa/dynamic-sliding-window/examples/LongestWithin.java
import java.util.List;

/**
 * The longest stretch whose total stays within a budget, in one pass.
 *
 * <p>Both search methods answer the same question. One tries every start
 * position and extends from it; the other moves two edges that never go
 * backward. Each reports how many values it looked at, because the difference
 * between them is the point.
 *
 * <p>{@code byWindow} assumes non-negative values. {@code byExhaustive} assumes
 * nothing and is far too slow to use, which is what makes it a reference.
 */
public final class LongestWithin {
  private LongestWithin() {}

  /** The longest width found, and how many values were examined. */
  public record Search(int width, int reads) {}

  /** Try each start position and extend from it. Re-reads what it already saw. */
  public static Search byEveryStart(List<Integer> values, int budget) {
    int best = 0;
    int reads = 0;

    for (int start = 0; start < values.size(); start++) {
      int total = 0;
      for (int end = start; end < values.size(); end++) {
        total += values.get(end);
        reads++;
        if (total > budget) {
          break;
        }
        best = Math.max(best, end - start + 1);
      }
    }

    return new Search(best, reads);
  }

  /** Move two edges, neither ever backward. One pass over the values. */
  public static Search byWindow(List<Integer> values, int budget) {
    int best = 0;
    int total = 0;
    int start = 0;
    int reads = 0;

    for (int end = 0; end < values.size(); end++) {
      total += values.get(end);
      reads++;

      // The front edge comes up only while the budget is broken, and stops as
      // soon as it holds. Both halves need the condition to be one-way.
      while (total > budget && start <= end) {
        total -= values.get(start);
        reads++;
        start++;
      }

      best = Math.max(best, end - start + 1);
    }

    return new Search(best, reads);
  }

  /**
   * Every stretch, with no early exit. Correct on any values, and far too slow
   * to use. It exists so the tests have something to be right against.
   */
  public static int byExhaustive(List<Integer> values, int budget) {
    int best = 0;

    for (int start = 0; start < values.size(); start++) {
      int total = 0;
      for (int end = start; end < values.size(); end++) {
        total += values.get(end);
        if (total <= budget) {
          best = Math.max(best, end - start + 1);
        }
      }
    }

    return best;
  }
}

File: lessons/dsa/dynamic-sliding-window/examples/longest_within.cpp
// The longest stretch whose total stays within a budget, in one pass.
//
// Both search functions answer the same question. One tries every start
// position and extends from it; the other moves two edges that never go
// backward. Each reports how many values it looked at, because the difference
// between them is the point.
//
// byWindow assumes non-negative values. byExhaustive assumes nothing and is far
// too slow to use, which is what makes it a reference.
#include <algorithm>
#include <vector>

// The longest width found, and how many values were examined.
struct Search {
  int width;
  int reads;
};

// Try each start position and extend from it. Re-reads what it already saw.
Search byEveryStart(const std::vector<int>& values, int budget) {
  int best = 0;
  int reads = 0;
  const int size = static_cast<int>(values.size());

  for (int start = 0; start < size; ++start) {
    int total = 0;
    for (int end = start; end < size; ++end) {
      total += values[static_cast<std::size_t>(end)];
      ++reads;
      if (total > budget) {
        break;
      }
      best = std::max(best, end - start + 1);
    }
  }

  return Search{best, reads};
}

// Move two edges, neither ever backward. One pass over the values.
Search byWindow(const std::vector<int>& values, int budget) {
  int best = 0;
  int total = 0;
  int start = 0;
  int reads = 0;
  const int size = static_cast<int>(values.size());

  for (int end = 0; end < size; ++end) {
    total += values[static_cast<std::size_t>(end)];
    ++reads;

    // The front edge comes up only while the budget is broken, and stops as
    // soon as it holds. Both halves need the condition to be one-way.
    while (total > budget && start <= end) {
      total -= values[static_cast<std::size_t>(start)];
      ++reads;
      ++start;
    }

    best = std::max(best, end - start + 1);
  }

  return Search{best, reads};
}

// Every stretch, with no early exit. Correct on any values, and far too slow to
// use. It exists so the tests have something to be right against.
int byExhaustive(const std::vector<int>& values, int budget) {
  int best = 0;
  const int size = static_cast<int>(values.size());

  for (int start = 0; start < size; ++start) {
    int total = 0;
    for (int end = start; end < size; ++end) {
      total += values[static_cast<std::size_t>(end)];
      if (total <= budget) {
        best = std::max(best, end - start + 1);
      }
    }
  }

  return best;
}
